app/controllers/spree/products_controller_decorator.rb

```diff
@@ -0,0 +1,36 @@
+Spree::ProductsController.class_eval do
+
+  after_action :create_product_view_event_on_intercom, only: :show, if: :spree_current_user
+  after_action :create_product_search_event_on_intercom, only: :index, if: [:spree_current_user, :searched?]
+
+  private
+
+    def create_product_view_event_on_intercom
```

Comment:
what we can do here is to create a concern and move this method there also converting this method into a generic one. like below.
```
def create_event_on_intercom
  Spree::Intercom::TrackEventsJob.perform_later("#{controller_name}_#{action_name}", send("#{action_name}_data"))
end
```

and then include this concern in your controllers. like below

```
 include XYZ

after_action :create_event_on_intercom, only: :show, if: :spree_current_user

def show_data
{...}
end